src/vs/workbench/api/common/extHostCommands.ts

```diff
@@ -247,20 +248,28 @@ export class CommandsConverter {
 
 		if (command.command && isNonEmptyArray(command.arguments)) {
 			// we have a contributed command with arguments. that
-			// means we don't want to send the arguments around
-
-			const id = ++this._cachIdPool;
-			this._cache.set(id, command);
-			disposables.add(toDisposable(() => {
-				this._cache.delete(id);
-				this._logService.trace('CommandsConverter#DISPOSE', id);
-			}));
-			result.$ident = id;
-
-			result.id = this._delegatingCommandId;
-			result.arguments = [id];
-
-			this._logService.trace('CommandsConverter#CREATE', command.command, id);
+			// means we don't want to send the arguments around unless
+			// we know exactly that the arguments are serializable
+
+			if (command.command === OpenAPICommand.ID || command.command === DiffAPICommand.ID) {
+				// this enables `vscode.open` and `vscode.diff` to benefit
+				// from the actual editor options to use based on the context
+				// (e.g. invocation from a custom tree view)
+				result.arguments = command.arguments;
```

Comment:
I have tried the same thing right now and I can report that it doesn't work... This is what now happens

* the arguments go through normal JSON-serialization (which for vscode.Range is special)
* the command get send to the renderer
* the renderer has `vscode.open` registered as well (need @alexdima to refresh my memory here)
* when selecting an item in the tree the renderer-side `vscode.open` command is invoked and it fails to [convert the selection/range argument](https://github.com/microsoft/vscode/blob/0b3aa0a6ea58a7cda4dfccd3322490b5241ef059/src/vs/workbench/api/common/apiCommands.ts#L123)

Reply:
It seems have coordinated well, I have pushed some work here: https://github.com/microsoft/vscode/compare/joh/open

The idea is that `CommandConverter#toInternal` doesn't only create/feed the delegate command but that it actually converts some (API) commands to the internal variant. E.g a tree item will then be using `_workbench.open` and so on

Reply:
@jrieken What can I help with?

Reply:
There is `src/vs/workbench/api/common/apiCommands.ts` which registers API commands on both sides - renderer and extension host. Not sure why that was done. Also, it registers commands that accept arguments which are types of `vsocde.d.ts` which means they "don't survive" the JSON serialisation/deserialisation, like `vscode.Selection`.

Reply:
I remember... We make API command registration extension host only so that we can multiple extension host (the `global` flag) but then we hit bugs with API-commands that escaped the extension host without treatment, like in markdown and then we added some API-commands to the renderer as well. It seems that some have evolved into a state where they only partially work in the renderer

Reply:
@jrieken I think your changes in https://github.com/microsoft/vscode/compare/joh/open are cleaner, so I am fine to drop my work and only keep the special massaging of the 2 commands on the main side. To clarify, this would mean we end up having `_workbench.open` as command in the tree and not `vscode.open`? I actually prefer that because now I need to depend on some internal API knowledge in the tree. I would rather depend on `_workbench.open` 👍

Reply:
Yeah, it will be `_workbench.open` and it's friends. I have created https://github.com/microsoft/vscode/pull/110455 which I am planning to merge despite some polish of where to revive uris and so

Reply:
Cool 👍

Reply:
https://github.com/microsoft/vscode/pull/110455 is merged